src/plonk/circuit.rs

```diff
@@ -460,14 +460,51 @@ impl<F> Mul<F> for Expression<F> {
 #[derive(Copy, Clone, Debug)]
 pub(crate) struct PointIndex(pub usize);
 
+/// A "virtual register" is a PLONK cell that has been queried at a particular relative
+/// offset within a custom gate.
+#[derive(Clone, Debug)]
+struct VirtualRegister {
+    column: Column<Any>,
+    rotation: Rotation,
+}
+
+impl<Col: Into<Column<Any>>> From<(Col, Rotation)> for VirtualRegister {
+    fn from((column, rotation): (Col, Rotation)) -> Self {
+        VirtualRegister {
+            column: column.into(),
+            rotation,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Gate<F: Field> {
+    name: &'static str,
+    poly: Expression<F>,
+    /// We track queried selectors separately from other registers, so that we can use
+    /// them to trigger debug checks on gates.
+    queried_selectors: Vec<VirtualRegister>,
+    queried_registers: Vec<VirtualRegister>,
```

Comment:
This might not be exactly how we need to store the selectors. What I want to do is:

- Track virtual registers (so we "know" the shape of the gate's tetris piece).
- Track which selectors are queried, at what relative offsets.
- Invert the `gate -> selectors` mapping into a `selector -> gates` mapping.
  - This might just be how we store it? Or maybe it's easier to just compute that in the dev tooling once we have all the gates.
- When a selector is enabled, look up the gates it was queried in. For each gate:
  - Use the selector's rotation to figure out where this instance of the gate is positioned.
  - Use the queried virtual registers to "mark" the corresponding absolute cells (or maybe the relative cells within a region) as "must be assigned to".
  - Return an error if not all marked cells are assigned to, with an error message indicating which gates are requiring the cell to be assigned.

Reply:
Another consideration: a gate could be toggled on `(1 - selector)`, i.e. only activated when the selector is NOT enable.

Reply:
I think we could better track that if instead of using `1 - selector` we used `!selector` somewhere? In the expression would be nice, but that would require `Expression` itself to do the tracking, which could get messy. And that doesn't cover more complex cases like `(2 - selector)`.

Reply:
By "could better track" I mean that `impl Neg for Selector` could update some map contained within `Selector` (either directly or by reference). That would be harder to do with `impl Sub<Selector> for Fp` which would need to interpret what that subtraction means. And in any case, the type of `selector` at this point is `Expression`.

Reply:
I've added the dev tooling for this in #274. As discussed in PR gardening, we'll leave this as-is for now (with `Selector` being a boolean selector that positively activates gates), and if we find we need negative-activation of gates or non-boolean selectors, we can add them.